### DISCUSSION TITLE
Expanding the rules of the Color Color pattern inside field initialization context

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#7225] [Language Ideas]
## Proposal

Expand the Color Color type binding in `Foo Bar = Bar.` initialization scenarios for fields

## Details

Currently, the Color Color pattern is well respected in most assignment contexts by attempting to bind to the type of the symbol instead of the instance, when failing to bind to the accessed member. For example, `Color Color = Color.Red;` and `Color = Color.Blue;` attempt to bind to the type named Color because Red and Blue are not valid instance members, and require a type qualification.

However, if the type of the symbol does not have the same name as the name of the symbol that is being assigned, this rule is not applied, meaning that code like the example below cannot work, despite having no reason to bind to the field itself upon its initialization.

## Example

Here is an [example](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoTMJgKYDOADhAMZ4AEAwgK4EAuA9mAHK6EnmYDemF/FBAGZBSCgGVmeKgBsIBAnwG8MAtYJEIALBQCyeegAtGAEyQAKAJRL1FFbdsBBevQBOUYDXp4AKhFcA5gYEFH6BwRQAvKH+QfQEAHSy8iFwMeHxCQCSMN6uAGZkeADcNuoAvmWVqgJlwqIUAPJGeK5l9g71pIwwDBTObh5evrERYXEh0ePBSXIKFGnTmTl5heSlNfzV1Zj1CCgAbA1Lihgd/F09fQPunt4n1HOT/S63wyezKRtql730L4M7iMMiEVq01pRojchvdRstcuCihtqkA) showing that the field is being accessed in its own initialization context, without attempting to bind to the type itself.

## Risks / Alternatives

The implementation cost is probably too high for the expected yield out of this improvement. It's not a very common encounter, and a solution is to avoid to name the symbol with the same name as the type that is being accessed in the assigned expression.

**Comment by Joe4evr** [comment]
Of note: currently this can be worked around by [qualifying the type name with `global::`](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQCYBGAWAChKYIxUBnABwgGNUACAYQFd6EB7MADlaDZm0oBvSu1nsCAZnlF2AZUGpOAGwj16MudIpyT8pQQAs7ALKoEAC34ATIgAoAlAdPsj37wEEEBExgbgRUABUIdABzO3p2KNj49gBedhitfmAILRAQJLiEegA6bV0EvAysnLyC6KLSgEkYcPQsVlQAbi9TAF9egeM5XsVldgB5Bwxe3z8xln4YPnZA4KhQ8MKU7eK06uzc/N3S8r12KszDupOSlraOth7h2SGhyjGCEgA2cZOpUZKRbLBCrIIhMKRBopM4JdJrCFbaHFMo6PTPEwLJYrBEbSEndj3DCPDjw8F4pHJFFE9qdZ5DIA=). I get that this might not look very elegant, and there might still be room here for something a little more pleasing, but *a* solution exists and given that it's pretty rare to encounter code like this to begin with, trying to solve it differently probably won't reach the bar.
